### ISSUE TITLE
Added search functionality for plugins (e.g. Spotify)

### ISSUE THREAD

**Issue by kaaholst** [kaaholst/android-squeezer#185]
Tested with spotify plugin.
The pull request fixes issue #151.

Original issue reported on github.com/nikclayton/android-squeezer by Flautz on 2016-12-16T14:18:32Z

**Comment by kaaholst** [comment]
Thank you!

I'd happily test this but I can't get Android Studio to build it at the moment. That's probably just because I've never used it before!!

Original issue reported on github.com/nikclayton/android-squeezer by elParaguayo on 2016-12-16T23:14:05Z

**Comment by kaaholst** [comment]
I had to change a couple of lines to get it to build but the search works very nicely. This makes a massive difference to me so thank you again.

Original issue reported on github.com/nikclayton/android-squeezer by elParaguayo on 2016-12-17T09:09:34Z

**Comment by kaaholst** [comment]
Thanks mate. I had to change two lines in SearchAdapter.java to get the head of branch "release-1.4.1" to build, too. Maybe we should open an issue for that...

Original issue reported on github.com/nikclayton/android-squeezer by Flautz on 2016-12-17T11:16:02Z

**Comment by kaaholst** [comment]
Might be better if you open the issue. I'm not an android/java person (in fact I've never coded in either, I just downloaded Android Studio because I wanted your pull request!). There's a good chance that my fix isn't the correct way of doing it!

Original issue reported on github.com/nikclayton/android-squeezer by elParaguayo on 2016-12-17T11:20:08Z

**Comment by kaaholst** [comment]
I am new to git, so this might be a dumb question.
But, why did you merge this into release 1.4.1, and not to the master?


Original issue reported on github.com/nikclayton/android-squeezer by Gerrelt on 2016-12-18T13:09:39Z

**Comment by kaaholst** [comment]
The master branch seemed to me to be a bit old since it points to a commit from June 2015. 

The develop branch seemed to be the right choice, but it is also a bit behind the "newest state". 

I think the right way would be to update the develop branch (and set it as default branch) and base pull requests on develop. Unfortunately we have no permissions to update develop... That's why I based my pull request on a release-branch - which is a dirty solution...

Original issue reported on github.com/nikclayton/android-squeezer by Flautz on 2016-12-18T15:10:51Z

**Comment by kaaholst** [comment]
OK, that's clear, I am going to switch to 1.4.1 too. Thanks.

Original issue reported on github.com/nikclayton/android-squeezer by Gerrelt on 2016-12-18T18:31:52Z

**Comment by kaaholst** [comment]
Is there any timescale for this reaching google play? Or perhaps someone can add the apk here.

Original issue reported on github.com/nikclayton/android-squeezer by ianmxyz on 2016-12-30T14:11:44Z

**Comment by kaaholst** [comment]
I've got a debug-apk for download here:
http://www.gerrelt.nl/RaspberryPi/squeezer/Squeezer-debug.apk
(you could also use [this QR code](http://images.gerrelt.nl/RaspberryPi/squeezer/squeezer_debug_1.4.2-beta-0.png) for easy downloading directly to your phone)

Instructions:
1. De-install currently installed Squeezer
2. In your phone's settings menu, somewhere under security, allow installing programs from unknown sources (you can switch it off again after installing Squeezer).
3. Download the .apk
4. Start and install the .apk by selecting it.

This is a beta test version, not an official release, let us know if you experience problems.
I can't create an official release, because I am not a member of this project. I really hope one of the members will find some time to create a new release. That would be excellent.


Original issue reported on github.com/nikclayton/android-squeezer by Gerrelt on 2016-12-30T16:29:59Z

**Comment by kaaholst** [comment]
I tried to contact the owner of the project about two weeks ago but did not receive a response. I also hope that the app will get new releases in the future... 
Maybe we should think about creating a fork if the author is not interested in the project anymore. Gerrelt Interested? :-) 

Original issue reported on github.com/nikclayton/android-squeezer by Flautz on 2017-01-01T20:57:04Z

**Comment by kaaholst** [comment]
Gerrelt many thanks, it seems to work well, there are few strange things that I will try to document. Nothing really to do with this app, but Spotify playlists are taking a long time to load on LMS at the moment and some seem to have vanished such as Discover Weekly. Is anyone else experiencing this?


Original issue reported on github.com/nikclayton/android-squeezer by ianmxyz on 2017-01-04T17:24:57Z

**Comment by kaaholst** [comment]
Thank you for these contributions,

I will talk with Nik and see if we can get a release with these and other contributions published soon.

Meanwhile, there are a few things that I would like to discuss with this pull request.

As the call to the SearchPluginDialog is in orderPage, it means that the dialog opens each time a page is downloaded. It happens if the search yields more than 20 results.

Additionally, the current solution for xmlbrowser_search type items, works with a search field embedded in the result list activity. The solution should preferably be the same for the two types of searchable items.

To use the embedded search field it can, for example. be implmented as follows.

Creating this method:
```java
private boolean isSearchable() {
    return plugin.isSearchable() || (parent != null && "search".equals(parent.getType()));
}
```
and use isSearchable() each time plugin.isSearchable() was called.

Addidionally it is necessary to change orderPage as follows:
```java
    if (!(isSearchable() && TextUtils.isEmpty(search)))
        service.pluginItems(start, plugin, parent, search, this);
```

This is because the plugin (at least the radiotime plugin) requires the search parameter. The xmlbrowser_search type plugins just return an empty response (count=0 and no title), which we currently rely on.

If this solution is chosen there is no need for the SearchPluginDialog.

In the other hand I kind of like the search dialog. If this solution is chosen it should also be used for xmlbrowser_search type plugins and the embedded search field shall be removed from the code. Also the call of the dialog shall not be in orderPage or be conditional. Furthermore with the dialog solution, it is either necessary to re-enter the activity to change the search criteria, or add a button/menu item to bring up the search dialog.

What do you think?

Thanks again for these contributions, they provide much needed/requested fuctionality for the Squeezer app.

Regards Kurt


Original issue reported on github.com/nikclayton/android-squeezer by kaaholst on 2017-01-06T16:03:10Z

**Comment by kaaholst** [comment]
kaaholst : Thanks for replying and it would be great if you could contact Nik.
My opinion on the search field: I've never seen the embedded search field, but I have no strong feelings about the embedded one or the dialog one Flautz added. I think both have their merits. Allthough the dialog one does make it very clear to the user what he/she has to do.
I've just tested, and I can confirm that the dialog search box re-appears when you scroll through search results bigger then 20. It would be nice if we could fix that.

ianmxyz : If you search for the word "discover" and then choose the "PlayLists" in the result view, then playlist "Discover Weekly" is at the top of the list. It might be a bit of a "workaround", but it does work.

Original issue reported on github.com/nikclayton/android-squeezer by Gerrelt on 2017-01-06T20:44:07Z

**Comment by kaaholst** [comment]
Gerrelt To see the embedded search field, select Internet Radio from the main page, and then select search. This works at least on my servers. Failing that implement as described above, the search field should appear instead of the dialog. Testet using My Apps / TuneIn Radio / Search TuneIn. I can't test with spotify as I don't have an account.

Flautz I forgot to mention one more thing. I don't like the "avoid permgen outofmemory errors" commit, because -XX:MaxPermSize=512m is a non standard option. Could you instead use another JVM which does not suffer from  permanent generation issues, e.g. the openjdk which is bundled with recent Android Studio versions.


Original issue reported on github.com/nikclayton/android-squeezer by kaaholst on 2017-01-06T21:25:36Z

**Comment by kaaholst** [comment]
kaaholst : Yes, I found it. I never noticed there was a search function for internet radio, pretty handy. That search field works nice too. It's handy to have the search field always at the top, after searching for something. It makes entering a new search word easier.

Original issue reported on github.com/nikclayton/android-squeezer by Gerrelt on 2017-01-07T10:43:34Z

**Comment by kaaholst** [comment]
kaaholst 
Thank you for your response. I checked the embedded search you proposed and like it very much. I did not now about this possibility and think we should prefer this. 

It seems that you already have the solution, I can just close my pull request and you add a new one with the embedded search style implementation?

Additionally, I agree with your bad feelings regarding the MaxPermSize commit. I added it by accident to this pull request and found a different solution for my build server in the mean time. 

Original issue reported on github.com/nikclayton/android-squeezer by Flautz on 2017-01-09T19:17:00Z

**Comment by kaaholst** [comment]
Yes I did test the embedded solution. It is on my [develop](https://github.com/kaaholst/android-squeezer/tree/develop) branch, which I have just pushed, so you can check it out if you like. It is actually build on your original commit, so credits can go where they belong.
I integrated pull requests 183-187 onto that branch, so we can test if it all works together.

Original issue reported on github.com/nikclayton/android-squeezer by kaaholst on 2017-01-10T07:28:26Z

**Comment by kaaholst** [comment]
kaaholst : I've tested the embedded solution, and it works great! I think I like it more then the dialog version too.

ianmxyz & All: the new version of the .apk is available on the same download location I mentioned (together with installation instructions) in an older comment of mine in this "thread".

Original issue reported on github.com/nikclayton/android-squeezer by Gerrelt on 2017-01-14T21:12:41Z

**Comment by kaaholst** [comment]
I've switched to the official Spotify plugin now as I was having trouble with Triode's plugin (list of playlists taking a long time to load, missing Discover Weekly). I can confirm that the search works fine either way now. One other inconsistency as follows:
When a track is found by a search a long touch will display a play menu, a short touch will give options to Star the track, add to a playlist, title radio, or get further tracks from the artist or album. This is great, however it would be even better to be able to access this menu from the playing track, which is currently not possible either directly or via the current playlist display. Hope this makes sense!
In essence it's great to be able to save the track to a Spotify playlist while it's playing.

Original issue reported on github.com/nikclayton/android-squeezer by ianmxyz on 2017-01-15T17:58:36Z

**Comment by kaaholst** [comment]
Gerrelt, thanks for your patched apk.

FYI, your patch fixes a similar issue on the mysqueezebox.com Tidal app. On 1.4.1 it was not possible to search, so Tidal was essentially unusable. So far things are working fine with your build.

Original issue reported on github.com/nikclayton/android-squeezer by mervinbeng on 2017-02-02T09:19:35Z

**Comment by kaaholst** [comment]
Hi -- would you mind splitting this in to two pull requests, one for the Jenkins functionality, and one for the search changes?

Thanks.

Original issue reported on github.com/nikclayton/android-squeezer by nikclayton on 2017-02-02T23:07:02Z

**Comment by kaaholst** [comment]
Sorry for the confusion, I accidentally chose my develop-branch instead of a separate branch for the pull request. I reorganized the commits and removed all commits which had nothing to do with the spotify search - now it should be OK.

Original issue reported on github.com/nikclayton/android-squeezer by Flautz on 2017-02-04T15:45:25Z

**Comment by kaaholst** [comment]
Thanks. I've merged this on to the `develop` branch.

Original issue reported on github.com/nikclayton/android-squeezer by nikclayton on 2017-02-05T22:09:04Z

**Comment by kaaholst** [comment]
I have squeezer 1.4.1 on android 5.1.1 (samsung), and still get the issue with Deezer.
Thanks.

Original issue reported on github.com/nikclayton/android-squeezer by assurancetourix on 2017-03-11T19:58:41Z

**Comment by kaaholst** [comment]
I've just released this in the first beta of 1.5.0. You can sign up to receive the beta versions at https://github.com/nikclayton/android-squeezer/wiki/Beta-testing.

Original issue reported on github.com/nikclayton/android-squeezer by nikclayton on 2017-03-20T20:42:07Z
